internal/controller/controllers/agent_controller.go

```diff
@@ -312,41 +312,46 @@ func (r *AgentReconciler) spokeKubeClient(ctx context.Context, clusterRef *aiv1b
 
 // Attempt to approve CSRs for agent. If already approved then the node will be marked as done
 // requeue means that approval will be attempted again
-func (r *AgentReconciler) tryApproveDay2CSRs(ctx context.Context, agent *aiv1beta1.Agent) bool {
+func (r *AgentReconciler) tryApproveDay2CSRs(ctx context.Context, agent *aiv1beta1.Agent, node *corev1.Node) bool {
 	r.Log.Infof("Approving CSRs for agent %s/%s", agent.Namespace, agent.Name)
-
-	hostname := getAgentHostname(agent)
 	var (
 		validateNodeCsr       nodeCsrValidator
 		shouldApproveMoreCSRs bool
 	)
 
-	clients, err := r.spokeKubeClient(ctx, agent.Spec.ClusterDeploymentName)
-	if err != nil {
-		r.Log.WithError(err).Errorf("Agent %s/%s: Failed to create spoke client", agent.Namespace, agent.Name)
-		return false
-	}
-
-	// TODO: Node name might be FQDN and not just host name if cluster is IPv6
-	node, err := clients.GetNode(hostname)
-	if err != nil {
-		if !k8serrors.IsNotFound(err) {
-			r.Log.WithError(err).Errorf("Agent %s/%s: Failed to get node %s", agent.Namespace, agent.Name, hostname)
-			return false
-		}
+	if node == nil {
 		validateNodeCsr = validateNodeClientCSR
 		shouldApproveMoreCSRs = true
 	} else {
 		validateNodeCsr = createNodeServerCsrValidator(node)
 		shouldApproveMoreCSRs = r.shouldApproveMoreCSRs(node)
 	}
 
+	clients, err := r.spokeKubeClient(ctx, agent.Spec.ClusterDeploymentName)
+	if err != nil {
+		r.Log.WithError(err).Errorf("Agent %s/%s: Failed to create spoke client", agent.Namespace, agent.Name)
+		return false
+	}
+
 	// Even if node is already ready, we try approving last time
 	r.approveAIHostsCSRs(clients, agent, validateNodeCsr)
 
 	return !shouldApproveMoreCSRs
 }
 
+func (r *AgentReconciler) getNode(ctx context.Context, agent *aiv1beta1.Agent) (*corev1.Node, error) {
+	hostname := getAgentHostname(agent)
+
+	clients, err := r.spokeKubeClient(ctx, agent.Spec.ClusterDeploymentName)
```

Comment:
Can you try to reduce the number of times you create a client?

I think it would be better to create it earlier and pass the reference around.